src/dmd/backend/inliner.d

```diff
@@ -0,0 +1,556 @@
+
+    /* If forward referencing an inline function, we'll have to
+     * write out the function when it eventually is defined
+     */
+    if (!sfunc.Sfunc) // this can happen for rtlsym functions
+    {
+    }
+    else if (sfunc.Sfunc.Fstartblock == null)
+        {   } //nwc_mustwrite(sfunc);
+    else
+    {   func_t *f = sfunc.Sfunc;
+
+        /* Check to see if we inline expand the function, or queue  */
+        /* it to be output.                                         */
+        if ((f.Fflags & (Finline | Finlinenest)) == Finline)
+            e = inline_expand(e,sfunc);
+        else
+            {   } //queue_func(sfunc);
+    }
+
+    return e;
+}
+
+/**********************************
+ * Inline expand a function.
+ * Params:
+ *      e = the OPcall or OPucall that calls sfunc, this gets free'd
+ *      sfunc = function being called that gets inlined
+ * Returns:
+ *      the expression replacing the function call
+ */
+
+private elem* inline_expand(elem *e,Symbol *sfunc)
+{
+    if (log) printf("inline_expand(e = %p, func %p = '%s')\n", e, sfunc, prettyident(sfunc));
+    //printf("before:\n"); elem_print(e);
+    //symbol_debug(sfunc);
+    assert(e.Eoper == OPcall || e.Eoper == OPucall);
+    func_t* f = sfunc.Sfunc;
+
+    // Declare all of sfunc's local symbols as symbols in globsym
+    const sistart = globsym.length;                      // where func's local symbols start
+    foreach (s; f.Flocsym[])
+    {
+        assert(s);
+        //if (!s)
+        //    continue;
+        //symbol_debug(s);
+        auto sc = s.Sclass;
+        switch (sc)
+        {
+            case SCparameter:
+            case SCfastpar:
+                sc = SCauto;
+                goto L1;
+            case SCregpar:
+                sc = SCregister;
+                goto L1;
+            case SCregister:
+            case SCauto:
+            case SCpseudo:
+            L1:
+            {
+                Symbol* snew = symbol_copy(s);
+                snew.Sclass = sc;
+                snew.Sfl = FLauto;
+                snew.Sflags |= SFLfree;
+                snew.Srange = null;
+                s.Sflags |= SFLreplace;
+                s.Ssymnum = symbol_add(snew);
+                break;
+            }
+            case SCglobal:
+            case SCstatic:
+                break;
+            default:
+                debug symbol_print(s);
+                assert(0);
+        }
+    }
+
+    static if (0)
+        foreach (i, s; globsym[])
+        {
+            if (i == sistart)
+                printf("---\n");
+            printf("[%d] %s\n", cast(int)i, s.Sident.ptr);
+        }
+
+    /* Create duplicate of function elems
+     */
+    assert(f.Fstartblock);
+    elem* ec = el_copytree(f.Fstartblock.Belem);
+
+    /* Walk the copied tree, replacing references to the old
+     * variables with references to the new
+     */
+    if (ec)
+    {
+        inline_expandwalk(ec);
+        if (config.flags3 & CFG3eh &&
```

Comment:
All compile-time asserts and segfaults look to be gone now based on testsuite run.

I see there's still some run-time corruption occurring with EH.
```d
// From runnable/sdtor.d
// Compile with -O -inline
struct C59
{
    void oops()
    {
        throw new Throwable("Oops!");
    }

    ~this()
    {
    }
}

void foo59()
{
    C59().oops();
}


void main()
{
    int i = 0;
    try
        foo59();  // segfault
    catch (Throwable)
        i = 1;
    assert(i == 1);
}
```